pkg/kubectl/cmd/config/config.go

```diff
@@ -66,59 +80,323 @@ func NewCmdConfig(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+					destinationFile = o.GetDefaultFilename()
+				}
+
+				configToWrite := getConfigFromFileOrDie(destinationFile)
+				delete(configToWrite.Clusters, key)
+
+				if err := clientcmd.WriteToFile(*configToWrite, destinationFile); err != nil {
+					return err
+				}
+			}
+		}
+
+		for key, context := range startingConfig.Contexts {
+			if _, exists := newConfig.Contexts[key]; !exists {
+				destinationFile := context.LocationOfOrigin
+				if len(destinationFile) == 0 {
+					destinationFile = o.GetDefaultFilename()
+				}
+
+				configToWrite := getConfigFromFileOrDie(destinationFile)
+				delete(configToWrite.Contexts, key)
+
+				if err := clientcmd.WriteToFile(*configToWrite, destinationFile); err != nil {
+					return err
+				}
+			}
+		}
+
+		for key, authInfo := range startingConfig.AuthInfos {
+			if _, exists := newConfig.AuthInfos[key]; !exists {
+				destinationFile := authInfo.LocationOfOrigin
+				if len(destinationFile) == 0 {
+					destinationFile = o.GetDefaultFilename()
+				}
+
+				configToWrite := getConfigFromFileOrDie(destinationFile)
+				delete(configToWrite.AuthInfos, key)
+
+				if err := clientcmd.WriteToFile(*configToWrite, destinationFile); err != nil {
+					return err
+				}
+			}
+		}
+
+	}
+
+	return nil
+}
+
+// writeCurrentContext takes three possible paths.
+// If newCurrentContext is the same as the startingConfig's current context, then we exit.
+// If newCurrentContext has a value, then that value is written into the default destination file.
+// If newCurrentConfig is empty, then we find the config file that is setting the CurrentContext and clear the value from that file
+func (o *PathOptions) writeCurrentContext(newCurrentContext string) error {
+	if startingConfig, err := o.getStartingConfig(); err != nil {
+		return err
+	} else if startingConfig.CurrentContext == newCurrentContext {
+		return nil
+	}
+
+	if len(newCurrentContext) > 0 {
+		destinationFile := o.GetDefaultFilename()
+		config := getConfigFromFileOrDie(destinationFile)
+		config.CurrentContext = newCurrentContext
+
+		if err := clientcmd.WriteToFile(*config, destinationFile); err != nil {
+			return err
+		}
+
+		return nil
 	}
 
-	if o.local {
-		if len(filename) > 0 {
-			return nil, "", fmt.Errorf("already loading from %v, cannot specify global as well", filename)
+	filesToCheck := make([]string, 0, len(o.LoadingRules.Precedence)+1)
+	filesToCheck = append(filesToCheck, o.LoadingRules.ExplicitPath)
+	filesToCheck = append(filesToCheck, o.LoadingRules.Precedence...)
+
+	for _, file := range filesToCheck {
+		currConfig := getConfigFromFileOrDie(file)
+
+		if len(currConfig.CurrentContext) > 0 {
+			currConfig.CurrentContext = newCurrentContext
+			if err := clientcmd.WriteToFile(*currConfig, file); err != nil {
+				return err
+			}
+
+			return nil
 		}
+	}
 
-		filename = ".kubeconfig"
-		config = getConfigFromFileOrDie(filename)
+	return nil
+}
 
+func (o *PathOptions) writePreferences(newPrefs clientcmdapi.Preferences) error {
```

Comment:
I think that better determination of where to write stanzas will be beneficial regardless of which path we take forward.  Rather than trying to significantly change our format right now, this is simply applies that detection and determination to the file we have now.